File: go/test/ruby_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cloudsmith-sdk/go"
	"github.com/voxgig-sdk/cloudsmith-sdk/go/core"

	vs "github.com/voxgig-sdk/cloudsmith-sdk/go/utility/struct"
)

func TestRubyEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Ruby(nil)
		if ent == nil {
			t.Fatal("expected non-nil RubyEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := rubyBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "ruby." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_RUBY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		rubyRef01Ent := client.Ruby(nil)
		rubyRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "ruby"}, setup.data), "ruby_ref01"))
		rubyRef01Data["identifier"] = setup.idmap["identifier01"]
		rubyRef01Data["owner"] = setup.idmap["owner01"]

		rubyRef01DataResult, err := rubyRef01Ent.Create(rubyRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		rubyRef01Data = core.ToMapAny(rubyRef01DataResult)
		if rubyRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		rubyRef01Match := map[string]any{
			"identifier": setup.idmap["identifier01"],
			"owner": setup.idmap["owner01"],
		}

		rubyRef01ListResult, err := rubyRef01Ent.List(rubyRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		rubyRef01List, rubyRef01ListOk := rubyRef01ListResult.([]any)
		if !rubyRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", rubyRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(rubyRef01List), map[string]any{"id": rubyRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		rubyRef01DataUp0Up := map[string]any{
			"identifier": setup.idmap["identifier"],
			"owner": setup.idmap["owner"],
		}

		rubyRef01MarkdefUp0Name := "auth_mode"
		rubyRef01MarkdefUp0Value := fmt.Sprintf("Mark01-ruby_ref01_%d", setup.now)
		rubyRef01DataUp0Up[rubyRef01MarkdefUp0Name] = rubyRef01MarkdefUp0Value

		rubyRef01ResdataUp0Result, err := rubyRef01Ent.Update(rubyRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		rubyRef01ResdataUp0 := core.ToMapAny(rubyRef01ResdataUp0Result)
		if rubyRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if rubyRef01ResdataUp0[rubyRef01MarkdefUp0Name] != rubyRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", rubyRef01MarkdefUp0Name, rubyRef01ResdataUp0[rubyRef01MarkdefUp0Name])
		}

		// LOAD
		rubyRef01MatchDt0 := map[string]any{}
		rubyRef01DataDt0Loaded, err := rubyRef01Ent.Load(rubyRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if rubyRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func rubyBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "ruby", "RubyTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read ruby test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse ruby test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"ruby01", "ruby02", "ruby03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CLOUDSMITH_TEST_RUBY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CLOUDSMITH_TEST_RUBY_ENTID": idmap,
		"CLOUDSMITH_TEST_LIVE":      "FALSE",
		"CLOUDSMITH_TEST_EXPLAIN":   "FALSE",
		"CLOUDSMITH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CLOUDSMITH_TEST_RUBY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add identifier alias for update test.
	if idmapResolved["identifier"] == nil {
		idmapResolved["identifier"] = idmapResolved["identifier01"]
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}

	if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CLOUDSMITH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCloudsmithSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/ruby_entity_test.rb
# Ruby entity test

require "minitest/autorun"
require "json"
require_relative "../Cloudsmith_sdk"
require_relative "runner"

class RubyEntityTest < Minitest::Test
  def test_create_instance
    testsdk = CloudsmithSDK.test(nil, nil)
    ent = testsdk.Ruby(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = ruby_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "ruby." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_RUBY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    ruby_ref01_ent = client.Ruby(nil)
    ruby_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.ruby"), "ruby_ref01"))
    ruby_ref01_data["identifier"] = setup[:idmap]["identifier01"]
    ruby_ref01_data["owner"] = setup[:idmap]["owner01"]

    ruby_ref01_data_result = ruby_ref01_ent.create(ruby_ref01_data, nil)
    ruby_ref01_data = Helpers.to_map(ruby_ref01_data_result)
    assert !ruby_ref01_data.nil?

    # LIST
    ruby_ref01_match = {
      "identifier" => setup[:idmap]["identifier01"],
      "owner" => setup[:idmap]["owner01"],
    }

    ruby_ref01_list_result = ruby_ref01_ent.list(ruby_ref01_match, nil)
    assert ruby_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(ruby_ref01_list_result),
      { "id" => ruby_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    ruby_ref01_data_up0_up = {
      "identifier" => setup[:idmap]["identifier"],
      "owner" => setup[:idmap]["owner"],
    }

    ruby_ref01_markdef_up0_name = "auth_mode"
    ruby_ref01_markdef_up0_value = "Mark01-ruby_ref01_#{setup[:now]}"
    ruby_ref01_data_up0_up[ruby_ref01_markdef_up0_name] = ruby_ref01_markdef_up0_value

    ruby_ref01_resdata_up0_result = ruby_ref01_ent.update(ruby_ref01_data_up0_up, nil)
    ruby_ref01_resdata_up0 = Helpers.to_map(ruby_ref01_resdata_up0_result)
    assert !ruby_ref01_resdata_up0.nil?
    assert_equal ruby_ref01_resdata_up0[ruby_ref01_markdef_up0_name], ruby_ref01_markdef_up0_value

    # LOAD
    ruby_ref01_match_dt0 = {}
    ruby_ref01_data_dt0_loaded = ruby_ref01_ent.load(ruby_ref01_match_dt0, nil)
    assert !ruby_ref01_data_dt0_loaded.nil?

  end
end

def ruby_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "ruby", "RubyTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = CloudsmithSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["ruby01", "ruby02", "ruby03", "package01", "package02", "package03", "repo01", "repo02", "repo03", "identifier01", "owner01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["CLOUDSMITH_TEST_RUBY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "CLOUDSMITH_TEST_RUBY_ENTID" => idmap,
    "CLOUDSMITH_TEST_LIVE" => "FALSE",
    "CLOUDSMITH_TEST_EXPLAIN" => "FALSE",
    "CLOUDSMITH_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["CLOUDSMITH_TEST_RUBY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end
  if idmap_resolved["identifier"].nil?
    idmap_resolved["identifier"] = idmap_resolved["identifier01"]
  end
  if idmap_resolved["owner"].nil?
    idmap_resolved["owner"] = idmap_resolved["owner01"]
  end

  if env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["CLOUDSMITH_APIKEY"],
      },
      extra || {},
    ])
    client = CloudsmithSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
